SCons/Defaults.xml

```diff
@@ -142,18 +142,40 @@ to the dictionary item's corresponding value;
 if the value is
 <literal>None</literal>,
 then the name is defined without an explicit value.
-Note that the resulting flags are sorted by keyword
-to ensure that the order of the options on the
-command line is consistent each time
-&scons;
-is run.
+Note that to ensure build order will be stable,
+&SCons; sorts the flags generated from the dictionary,
+since a changed command line triggers a rebuild
+(older &Python; interpreters did
+not guarantee to preserve dictionary keys in insert order),
+This precaution may be removed in a future version.
 </para>
 
 <example_commands>
 # Will add -DA -DB=2 to POSIX compiler command lines,
 # and /DA /DB=2 to Microsoft Visual C++ command lines.
 env = Environment(CPPDEFINES={'B':2, 'A':None})
 </example_commands>
+
+<para>
+Depending on how contents are added to &cv-CPPDEFINES;,
+it may end up as a compound type,
+for example a list containing strings, tuples and/or dictionaries.
+&SCons; will still expand this correctly.
+</para>
+
+<para>
+Note that if a value contains spaces (or other syntax that
+will be meaningful to the shell that calls the compiler)
+it needs to be quoted:
+</para>
+
+<example_commands>
+# SPACED will just get the string 'spaced', arg will be a separate arg
+env = Environment(CPPDEFINES={'SPACED': 'spaced arg'})
```

Comment:
Is that the current functionality? Hmm. not what I'd expect.

Reply:
Revisited this. At the core, this is not an SCons limitation (although it becomes one because we always invoke a shell to run the compile command), but the fact that the definition is already in a Python string makes it really hard to get right.  Here's what the cpp manpage says:

> If you are invoking the preprocessor from a shell or shell-like program you may need to use the shell's quoting syntax to protect characters such as spaces that have a meaning in the shell syntax.

GCC's cpp supports function-like macros on the command line, MSVC does not. They have the extra problem of using parens, which matter to the shell also.

On the whole, we'd need to emit options this way:
```
gcc -D"FOO=thing with strings"
```
That is, quoting the whole definition rather than the right-hand-side, which I think there is no way to do with current syntax - we'd have to hack on `_defines` I assume.  Open to suggestions on how to describe this, if we should at all.  I'm pushing a new revision which removes the example, because the example (just invented by me, not tested at the time) doesn't actually work - take a look at whether that wording covers the case or if we should just "remain silent" on this topic and figure people will only use simple replacement in CPPDEFINES.

Reply:
adding MSVC's manpage on the function-like macro topic for completeness:

> The `/D` option doesn't support function-like macro definitions. To insert definitions that can't be defined on the command line, consider the `/FI (Name forced include file)` compiler option.

and cpp's:

> If you wish to define a function-like macro on the command line, write its argument list with surrounding parentheses before the equals sign (if any).  Parentheses are meaningful to most shells, so you should quote the option.  With sh and csh, `-D'name(args...)=definition'` works.